### PR TITLE
Enable heterogeneous porosity and permeability relations via mesh input.

### PR DESCRIPTION
**Pull request by TomFischer**
Use parameters within the porosity and permeability description.

ToDo:
- [x] clean commit history
- [x] add heterogeneous test case(s)
- [x] documentation
- [x] discuss form of input in project files.

The reading of the porosity and permeability now involves the reading of a parameter. As a consequence a complete project file has to be provided as xml code in the particular tests. For this reason, I removed the unit tests checking those functionality. If there are objections against this removal please discuss this here.

### CODE REVIEW

**Comment by waltherm** [comment]
How general is that implementation - is it only working for porosity and permeability or also for all other variables given to a process?

**Comment by TomFischer** [comment]
At the moment only porosity and permeability are working. So there is a lot of work open.

**Comment by TomFischer** [comment]
@wenqing Please seemy updated introduction to this PR.

File: MaterialLib/PorousMedium/Permeability/Permeability.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+{
+public:
+    explicit Permeability(
+        ProcessLib::Parameter<double> const& permeability_parameter,
+        int const dimension)
+        : _permeability_parameter(permeability_parameter), _dimension(dimension)
+    {
+        auto const number_of_components =
+            _permeability_parameter.getNumberOfComponents();
+
```
Comment (wenqing) [inline MaterialLib/PorousMedium/Permeability/Permeability.h]: How about to use the following lines to replace the long switch case?
```
if (number_of_components != (_dimension *_dimension))
{
     OGS_FATAL(
                "The given parameter has %d component, but the "
                "permeability tensor is defined for a %d dimensional "
                "problem.",
                        _dimension * _dimension, _dimension);
}
```

File: MaterialLib/PorousMedium/Permeability/createPermeabilityModel.cpp [hunk 2]
```diff
@@ -18,39 +18,59 @@
...
-            dim = 3;
-            break;
-        default:
+        auto const& permeability_parameter = ProcessLib::findParameter<double>(
+            config,
+            //! \ogs_file_param_special{material__porous_medium__permeability__permeability_tensor_entries}
+            "permeability_tensor_entries", parameters, 0);
+
+
+        int dimension = 0;
```
Comment (wenqing) [inline MaterialLib/PorousMedium/Permeability/createPermeabilityModel.cpp]: How about the following code?
```
const int dimension = static_cast_<int>(std::sqrt(permeability_parameter.getNumberOfComponents()));
if (permeability_parameter.getNumberOfComponents() % (dimension * dimension)  != 0)
{
 OGS_FATAL(
           "Number of values for permeability tensor must be 1, 4 or 9.");
}
```

File: MaterialLib/PorousMedium/Permeability/Permeability.h (hunk 1 above)
Comment (TomFischer) [reply MaterialLib/PorousMedium/Permeability/Permeability.h]: Both the current code and your suggested code handle only tensors that are dense, i.e., 1d - 1 component, 2d - 4 components and 3d - 9 components. I wrote this with the idea in mind to generalize this also to diagonal and symmetric tensor cases, i.e., 2d tensor that can have 2 diagonal components or 3 components in the symmetric case and 3d tensor with 3 or 6 components. I'll try to change the code here and in createPermeabilityModel.cpp. What do you think, should I do it now or after your PR #1970 is rebased, reviewed and merged?

Comment (wenqing) [reply MaterialLib/PorousMedium/Permeability/Permeability.h]: I see. 
Maybe the diagonal elements are only needed for input together with an optional input of system rotational angles for the anisotropic permeability in local system, e.g. permeability in a inclined bedding plane.  In the practice, there is seldom case that all elements of a permeability tensor are given or measured directly. The input of diagonal elements with an optional input of system rotational angles  sounds more practical for permeability, thermal conductivity or concentration diffusion tensor. We may need a general input and calculation (for possible system rotation)  interface for such tensors. So far, I am satisfied with your current implementation.

File: MaterialLib/PorousMedium/Permeability/Permeability.h [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            case 4:
+                if (_dimension != 2)
+                    OGS_FATAL(
+                        "The given parameter has 4 components, but the "
+                        "permeability tensor is defined for a %d dimensional "
+                        "problem.",
+                        _dimension);
+                break;
+            case 9:
+                if (_dimension != 3)
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Permeability/Permeability.h]: all cases could be handled shorter when comparing `number_of_components` to `_dimension` squared.

File: MaterialLib/PorousMedium/Permeability/createPermeabilityModel.cpp [hunk 4]
```diff
@@ -18,39 +18,59 @@
...
-Eigen::MatrixXd createPermeabilityModel(BaseLib::ConfigTree const& config)
+std::unique_ptr<Permeability> createPermeabilityModel(
+    BaseLib::ConfigTree const& config,
+    std::vector<std::unique_ptr<ProcessLib::ParameterBase>> const& parameters)
 {
-    auto const values =
-        //! \ogs_file_param{material__porous_medium__permeability__values}
-        config.getConfigParameter<std::vector<double>>("values");
+    //! \ogs_file_param{material__porous_medium__permeability}
+    auto const type = config.getConfigParameter<std::string>("type");
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Permeability/createPermeabilityModel.cpp]: Mismatch; ogs_file_param must check for "type". Maybe
```
\ogs_file_param{material__porous_medium__permeability__type}
```
?

File: MaterialLib/PorousMedium/Porosity/createPorosityModel.cpp [hunk 5]
```diff
@@ -15,25 +15,49 @@
...
+    std::vector<std::unique_ptr<ProcessLib::ParameterBase>> const& parameters)
 {
     //! \ogs_file_param{material__porous_medium__porosity__type}
     auto const type = config.getConfigParameter<std::string>("type");
 
     if (type == "Constant")
+    {
+        auto const& constant_porosity = ProcessLib::findParameter<double>(
+            config,
+            //! \ogs_file_param_special{material__porous_medium__porosity__}
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Porosity/createPorosityModel.cpp]: something is missing at the end...

File: MaterialLib/PorousMedium/Porosity/createPorosityModel.cpp [hunk 6]
```diff
@@ -15,25 +15,49 @@
...
             //! \ogs_file_param{material__porous_medium__porosity__Constant__value}
-            config.getConfigParameter<double>("value"));
+            constant_porosity);
+    }
+
+    if (type == "ElementBased")
+    {
+        auto const& element_based_porosity = ProcessLib::findParameter<double>(
+            config,
+            //! \ogs_file_param_special{material__porous_medium__porosity__}
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Porosity/createPorosityModel.cpp]: dto

File: MaterialLib/PorousMedium/Porosity/createPorosityModel.h [hunk 7]
```diff
@@ -27,9 +28,11 @@ class Porosity;
 
 /** Create a porosity model
  *  @param config  ConfigTree object has a tag of `<porosity>`
+ *  @param parameters ToDo TomFischer
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Porosity/createPorosityModel.h]: A short note on what is todo would be helpful.

**Comment by endJunction** [comment]
Tests/Data changes should be dropped now as git lfs is working.

**Comment by wenqing** [comment]
:+1: . I am going to rebase.

**Pull request merged**